File: clients/python/src/pawrly/transports/convert.py
"""Build result dataclasses from the REST JSON shapes (snake_case, 1:1 with the
engine's serde types)."""

import re

from ..result import (
    CacheEntryInfo,
    CatalogSnapshot,
    ColumnSpec,
    FunctionArg,
    FunctionColumn,
    FunctionDescription,
    FunctionInfo,
    RefreshCatalogOutcome,
    RefreshOutcome,
    ReloadReport,
    SchemaSummary,
    SemanticDimension,
    SemanticFilter,
    SemanticMeasure,
    SemanticModelDescription,
    SemanticModelInfo,
    SemanticRelationship,
    SemanticSegment,
    SourceInfo,
    SourceTestReport,
    TableDescription,
    TableInfo,
    TableName,
    TableSummary,
    VacuumReport,
)

_HUMANTIME_MS = {
    "ns": 1e-6,
    "us": 1e-3,
    "µs": 1e-3,
    "ms": 1.0,
    "s": 1000.0,
    "m": 60_000.0,
    "h": 3_600_000.0,
    "d": 86_400_000.0,
}


def humantime_ms(value: str | None) -> float:
    """Parse a humantime duration (`"1s 500ms"`, `"150ms"`) into milliseconds."""
    if not value:
        return 0.0
    total = 0.0
    for num, unit in re.findall(r"(\d+(?:\.\d+)?)\s*(ns|us|µs|ms|s|m|h|d)", value):
        total += float(num) * _HUMANTIME_MS[unit]
    return total


def table_name(d: dict) -> TableName:
    d = d or {}
    return TableName(schema=d.get("schema", ""), table=d.get("table", ""))


def source_info(d: dict) -> SourceInfo:
    return SourceInfo(
        name=d.get("name", ""),
        kind=d.get("kind", ""),
        status=d.get("status", ""),
        status_detail=d.get("status_detail"),
        sub_kind=d.get("sub_kind"),
        table_count=d.get("table_count", 0),
        registered_at=d.get("registered_at", ""),
    )


def table_info(d: dict) -> TableInfo:
    return TableInfo(
        name=table_name(d.get("name")),
        kind=d.get("kind", ""),
        description=d.get("description"),
        row_count_estimate=d.get("row_count_estimate"),
        cached=d.get("cached", False),
        required_filters=d.get("required_filters", []),
    )


def column_spec(d: dict) -> ColumnSpec:
    return ColumnSpec(
        name=d.get("name", ""),
        data_type=d.get("data_type", ""),
        nullable=d.get("nullable", False),
        description=d.get("description"),
        is_filter_pushable=d.get("is_filter_pushable", False),
        is_required_filter=d.get("is_required_filter", False),
    )


def table_description(d: dict) -> TableDescription:
    return TableDescription(
        table=table_info(d.get("table")),
        columns=[column_spec(c) for c in d.get("columns", [])],
        pushable_filter_columns=d.get("pushable_filter_columns", []),
        examples=d.get("examples", []),
        wiki=d.get("wiki"),
    )


def catalog_snapshot(d: dict) -> CatalogSnapshot:
    return CatalogSnapshot(
        schemas=[
            SchemaSummary(
                name=s.get("name", ""),
                kind=s.get("kind", ""),
                tables=[
                    TableSummary(
                        name=t.get("name", ""),
                        columns=t.get("columns", ""),
                        required_filters=t.get("required_filters", []),
                    )
                    for t in s.get("tables", [])
                ],
            )
            for s in d.get("schemas", [])
        ]
    )


def cache_entry(d: dict) -> CacheEntryInfo:
    return CacheEntryInfo(
        name=table_name(d.get("name")),
        mode=d.get("mode", ""),
        written_at=d.get("written_at", ""),
        expires_at=d.get("expires_at"),
        row_count=d.get("row_count", 0),
        size_bytes=d.get("size_bytes", 0),
        file_count=d.get("file_count", 0),
    )


def function_info(d: dict) -> FunctionInfo:
    return FunctionInfo(
        namespace=d.get("namespace", ""),
        name=d.get("name", ""),
        kind=d.get("kind", ""),
        builtin=d.get("builtin", False),
        signature=d.get("signature", ""),
        description=d.get("description"),
    )


def function_description(d: dict) -> FunctionDescription:
    return FunctionDescription(
        namespace=d.get("namespace", ""),
        name=d.get("name", ""),
        kind=d.get("kind", ""),
        builtin=d.get("builtin", False),
        signature=d.get("signature", ""),
        description=d.get("description"),
        wiki=d.get("wiki"),
        examples=d.get("examples", []),
        args=[
            FunctionArg(
                name=a.get("name", ""),
                type=a.get("type", ""),
                required=a.get("required", False),
                default=a.get("default"),
                description=a.get("description"),
                tool_arg=a.get("tool_arg"),
            )
            for a in d.get("args", [])
        ],
        returns=[
            FunctionColumn(
                name=c.get("name", ""),
                type=c.get("type", ""),
                source=c.get("source"),
                description=c.get("description"),
            )
            for c in d.get("returns", [])
        ],
    )


def semantic_model_info(d: dict) -> SemanticModelInfo:
    return SemanticModelInfo(
        name=d.get("name", ""),
        description=d.get("description"),
        source=d.get("source", ""),
        dimension_count=d.get("dimension_count", 0),
        measure_count=d.get("measure_count", 0),
    )


def source_test_report(d: dict) -> SourceTestReport:
    return SourceTestReport(
        name=d.get("name", ""),
        ok=d.get("ok", False),
        latency_ms=humantime_ms(d.get("latency")),
        detail=d.get("detail"),
    )


def reload_report(d: dict) -> ReloadReport:
    return ReloadReport(
        sources_added=d.get("sources_added", 0),
        sources_removed=d.get("sources_removed", 0),
        sources_changed=d.get("sources_changed", 0),
    )


def refresh_catalog_outcome(d: dict) -> RefreshCatalogOutcome:
    return RefreshCatalogOutcome(
        sources_refreshed=d.get("sources_refreshed", 0),
        tables_discovered=d.get("tables_discovered", 0),
    )


def refresh_outcome(d: dict) -> RefreshOutcome:
    return RefreshOutcome(
        table=table_name(d.get("table")),
        rows_written=d.get("rows_written", 0),
        size_bytes=d.get("size_bytes", 0),
        elapsed_ms=humantime_ms(d.get("elapsed")),
        expires_at=d.get("expires_at"),
    )


def vacuum_report(d: dict) -> VacuumReport:
    return VacuumReport(
        entries_removed=d.get("entries_removed", 0),
        files_removed=d.get("files_removed", 0),
        bytes_reclaimed=d.get("bytes_reclaimed", 0),
    )


def _measure_agg(agg) -> tuple[str, str | None]:
    # Unit variants serialize as a bare string; `Custom` as `{custom: {sql}}`.
    if isinstance(agg, str):
        return agg, None
    if isinstance(agg, dict) and "custom" in agg:
        return "custom", (agg["custom"] or {}).get("sql")
    return "", None


def semantic_model_description(d: dict) -> SemanticModelDescription:
    return SemanticModelDescription(
        name=d.get("name", ""),
        description=d.get("description"),
        source=d.get("source", ""),
        primary_key=d.get("primary_key", []),
        dimensions=[
            SemanticDimension(
                name=x.get("name", ""),
                expr=x.get("expr", ""),
                type=x.get("type", ""),
                grains=x.get("grains", []),
                description=x.get("description"),
            )
            for x in d.get("dimensions", [])
        ],
        measures=[_measure(x) for x in d.get("measures", [])],
        relationships=[
            SemanticRelationship(
                name=x.get("name", ""),
                kind=x.get("kind", ""),
                target=x.get("target", ""),
                on=x.get("on", ""),
            )
            for x in d.get("relationships", [])
        ],
        segments=[
            SemanticSegment(
                name=x.get("name", ""),
                description=x.get("description"),
                filters=[
                    SemanticFilter(
                        member=f.get("member", ""),
                        op=f.get("op", ""),
                        values=f.get("values"),
                    )
                    for f in x.get("filters", [])
                ],
            )
            for x in d.get("segments", [])
        ],
    )


def _measure(x: dict) -> SemanticMeasure:
    agg, custom_sql = _measure_agg(x.get("agg"))
    return SemanticMeasure(
        name=x.get("name", ""),
        agg=agg,
        custom_sql=custom_sql,
        expr=x.get("expr", ""),
        filters=x.get("filters", []),
        format=x.get("format"),
        description=x.get("description"),
    )

File: clients/typescript/src/transports/convert.ts
// Build result objects from the REST JSON shapes (snake_case, 1:1 with the
// engine's serde types) into the SDK's camelCase interfaces.
import type {
  CacheEntryInfo,
  CatalogSnapshot,
  ColumnSpec,
  FilterOp,
  FunctionArg,
  FunctionColumn,
  FunctionDescription,
  FunctionInfo,
  RefreshCatalogOutcome,
  RefreshOutcome,
  ReloadReport,
  SchemaSummary,
  SemanticDimension,
  SemanticMeasure,
  SemanticModelDescription,
  SemanticModelInfo,
  SemanticRelationship,
  SemanticSegment,
  SourceInfo,
  SourceTestReport,
  TableDescription,
  TableInfo,
  TableName,
  TableSummary,
  VacuumReport,
} from "../result.js";

type J = Record<string, unknown>;
const obj = (v: unknown): J => (v ?? {}) as J;
const arr = (v: unknown): J[] => (Array.isArray(v) ? (v as J[]) : []);
const str = (v: unknown): string | undefined => (typeof v === "string" ? v : undefined);
const int = (v: unknown): number => (typeof v === "number" ? v : 0);

const HUMANTIME_MS: Record<string, number> = {
  ns: 1e-6,
  us: 1e-3,
  "µs": 1e-3,
  ms: 1,
  s: 1000,
  m: 60_000,
  h: 3_600_000,
  d: 86_400_000,
};

/** Parse a humantime duration (`"1s 500ms"`, `"150ms"`) into milliseconds. */
export function humantimeMs(value: unknown): number {
  if (typeof value !== "string") return 0;
  let total = 0;
  for (const [, num, unit] of value.matchAll(/(\d+(?:\.\d+)?)\s*(ns|us|µs|ms|s|m|h|d)/g)) {
    total += parseFloat(num) * (HUMANTIME_MS[unit] ?? 0);
  }
  return total;
}

export function tableName(d: unknown): TableName {
  const o = obj(d);
  return { schema: (o.schema as string) ?? "", table: (o.table as string) ?? "" };
}

export function sourceInfo(d: J): SourceInfo {
  return {
    name: (d.name as string) ?? "",
    kind: (d.kind as string) ?? "",
    status: (d.status as string) ?? "",
    statusDetail: str(d.status_detail),
    subKind: str(d.sub_kind),
    tableCount: (d.table_count as number) ?? 0,
    registeredAt: (d.registered_at as string) ?? "",
  };
}

export function tableInfo(d: J): TableInfo {
  return {
    name: tableName(d.name),
    kind: (d.kind as string) ?? "",
    description: str(d.description),
    rowCountEstimate: d.row_count_estimate as number | undefined,
    cached: (d.cached as boolean) ?? false,
    requiredFilters: (d.required_filters as string[]) ?? [],
  };
}

export function columnSpec(d: J): ColumnSpec {
  return {
    name: (d.name as string) ?? "",
    dataType: (d.data_type as string) ?? "",
    nullable: (d.nullable as boolean) ?? false,
    description: str(d.description),
    isFilterPushable: (d.is_filter_pushable as boolean) ?? false,
    isRequiredFilter: (d.is_required_filter as boolean) ?? false,
  };
}

export function tableDescription(d: J): TableDescription {
  return {
    table: tableInfo(obj(d.table)),
    columns: arr(d.columns).map(columnSpec),
    pushableFilterColumns: (d.pushable_filter_columns as string[]) ?? [],
    examples: (d.examples as string[]) ?? [],
    wiki: str(d.wiki),
  };
}

export function catalogSnapshot(d: J): CatalogSnapshot {
  return {
    schemas: arr(d.schemas).map(
      (s): SchemaSummary => ({
        name: (s.name as string) ?? "",
        kind: (s.kind as string) ?? "",
        tables: arr(s.tables).map(
          (t): TableSummary => ({
            name: (t.name as string) ?? "",
            columns: (t.columns as string) ?? "",
            requiredFilters: (t.required_filters as string[]) ?? [],
          }),
        ),
      }),
    ),
  };
}

export function cacheEntry(d: J): CacheEntryInfo {
  return {
    name: tableName(d.name),
    mode: (d.mode as string) ?? "",
    writtenAt: (d.written_at as string) ?? "",
    expiresAt: str(d.expires_at),
    rowCount: (d.row_count as number) ?? 0,
    sizeBytes: (d.size_bytes as number) ?? 0,
    fileCount: (d.file_count as number) ?? 0,
  };
}

export function functionInfo(d: J): FunctionInfo {
  return {
    namespace: (d.namespace as string) ?? "",
    name: (d.name as string) ?? "",
    kind: (d.kind as string) ?? "",
    builtin: (d.builtin as boolean) ?? false,
    signature: (d.signature as string) ?? "",
    description: str(d.description),
  };
}

export function functionDescription(d: J): FunctionDescription {
  return {
    namespace: (d.namespace as string) ?? "",
    name: (d.name as string) ?? "",
    kind: (d.kind as string) ?? "",
    builtin: (d.builtin as boolean) ?? false,
    signature: (d.signature as string) ?? "",
    description: str(d.description),
    wiki: str(d.wiki),
    examples: (d.examples as string[]) ?? [],
    args: arr(d.args).map(
      (a): FunctionArg => ({
        name: (a.name as string) ?? "",
        type: (a.type as string) ?? "",
        required: (a.required as boolean) ?? false,
        default: str(a.default),
        description: str(a.description),
        toolArg: str(a.tool_arg),
      }),
    ),
    returns: arr(d.returns).map(
      (c): FunctionColumn => ({
        name: (c.name as string) ?? "",
        type: (c.type as string) ?? "",
        source: str(c.source),
        description: str(c.description),
      }),
    ),
  };
}

export function semanticModelInfo(d: J): SemanticModelInfo {
  return {
    name: (d.name as string) ?? "",
    description: str(d.description),
    source: (d.source as string) ?? "",
    dimensionCount: (d.dimension_count as number) ?? 0,
    measureCount: (d.measure_count as number) ?? 0,
  };
}

export function sourceTestReport(d: J): SourceTestReport {
  return {
    name: (d.name as string) ?? "",
    ok: (d.ok as boolean) ?? false,
    latencyMs: humantimeMs(d.latency),
    detail: str(d.detail),
  };
}

export function reloadReport(d: J): ReloadReport {
  return {
    sourcesAdded: int(d.sources_added),
    sourcesRemoved: int(d.sources_removed),
    sourcesChanged: int(d.sources_changed),
  };
}

export function refreshCatalogOutcome(d: J): RefreshCatalogOutcome {
  return {
    sourcesRefreshed: int(d.sources_refreshed),
    tablesDiscovered: int(d.tables_discovered),
  };
}

export function refreshOutcome(d: J): RefreshOutcome {
  return {
    table: tableName(d.table),
    rowsWritten: int(d.rows_written),
    sizeBytes: int(d.size_bytes),
    elapsedMs: humantimeMs(d.elapsed),
    expiresAt: str(d.expires_at),
  };
}

export function vacuumReport(d: J): VacuumReport {
  return {
    entriesRemoved: int(d.entries_removed),
    filesRemoved: int(d.files_removed),
    bytesReclaimed: int(d.bytes_reclaimed),
  };
}

/** MeasureAgg is a bare string for unit variants, or `{ custom: { sql } }`. */
function measureAgg(agg: unknown): { agg: string; customSql?: string } {
  if (typeof agg === "string") return { agg };
  if (agg && typeof agg === "object" && "custom" in agg) {
    return { agg: "custom", customSql: str((agg as J).custom && obj((agg as J).custom).sql) };
  }
  return { agg: "" };
}

export function semanticModelDescription(d: J): SemanticModelDescription {
  return {
    name: (d.name as string) ?? "",
    description: str(d.description),
    source: (d.source as string) ?? "",
    primaryKey: (d.primary_key as string[]) ?? [],
    dimensions: arr(d.dimensions).map(
      (x): SemanticDimension => ({
        name: (x.name as string) ?? "",
        expr: (x.expr as string) ?? "",
        type: (x.type as string) ?? "",
        grains: (x.grains as string[]) ?? [],
        description: str(x.description),
      }),
    ),
    measures: arr(d.measures).map((x): SemanticMeasure => {
      const { agg, customSql } = measureAgg(x.agg);
      return {
        name: (x.name as string) ?? "",
        agg,
        customSql,
        expr: (x.expr as string) ?? "",
        filters: (x.filters as string[]) ?? [],
        format: str(x.format),
        description: str(x.description),
      };
    }),
    relationships: arr(d.relationships).map(
      (x): SemanticRelationship => ({
        name: (x.name as string) ?? "",
        kind: (x.kind as string) ?? "",
        target: (x.target as string) ?? "",
        on: (x.on as string) ?? "",
      }),
    ),
    segments: arr(d.segments).map(
      (x): SemanticSegment => ({
        name: (x.name as string) ?? "",
        description: str(x.description),
        filters: arr(x.filters).map((f) => ({
          member: (f.member as string) ?? "",
          op: f.op as FilterOp,
          values: f.values as string[] | undefined,
        })),
      }),
    ),
  };
}
